Specialize sym node when used as device kwarg

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131811

Fixes https://github.com/pytorch/pytorch/issues/131189.

We specialize the symint in python_arg_parser when used as kwarg device.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ydwu4 committed:
[dynamo] specialize sym node when used as device kwarg

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] specialize sym node when used as device kwarg"


Fixes https://github.com/pytorch/pytorch/issues/131189.

By specializing the sym node when used as kwargs device.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented:
While it is possible that this is the best possible implementation, I am meh about it because it does not work with non strict export. Are we sure it's not possible to modify the argument parser to force the spec?

ydwu4 committed:
Update on "[dynamo] specialize sym node when used as device kwarg"


Fixes https://github.com/pytorch/pytorch/issues/131189.

By specializing the sym node when used as kwargs device.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 commented:
> Are we sure it's not possible to modify the argument parser to force the spec?

Updated the pr. In python_arg_parser.cpp, we allow the signature check to pass when device receives a SymInt input and 2. specialize it to a integer.

ydwu4 committed:
Update on "[dynamo] specialize sym node when used as device kwarg"


Fixes https://github.com/pytorch/pytorch/issues/131189.

We specialize the symint in python_arg_parser when used as kwarg device.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -1526,7 +1528,16 @@ bool FunctionSignature::parse(
+        // then cast it back to python.
+        auto guarded_int = py::cast<c10::SymInt>(py::handle(obj))
+                               .guard_int(__FILE__, __LINE__);
+        dst[i++] = py::cast(guarded_int).ptr();
```
This sure looks like it will result in an excessive decref later (if dst is owning). Can you argue why this has the right lifetime? (cc @albanD)

ydwu4 replied:
Yeah, the life time is wrong. I'm not very familar to pybind and digged a bit into it: dst doesn't own `py::cast(guarded_int)`. The integer is a local variable so dst will have a dangling pointer to it after we exits the if-statement. The previous diff works probably because gc hasn't cleaned the symint up. Thanks for pointing it out! 

Updated the diff so that when we're actually creating Device from argument, we do the specialization and create a device using integer index.

ezyang reviewed (changes requested):
see comment

ydwu4 committed:
Update on "Specialize sym node when used as device kwarg"


Fixes https://github.com/pytorch/pytorch/issues/131189.

We specialize the symint in python_arg_parser when used as kwarg device.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "Specialize sym node when used as device kwarg"


Fixes https://github.com/pytorch/pytorch/issues/131189.

We specialize the symint in python_arg_parser when used as kwarg device.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

albanD reviewed (approved):
Updated code sounds good

ydwu4 commented:
@pytorchbot merge
